test/parallel/test-assert-async.js

```diff
@@ -0,0 +1,98 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const {promisify} = require('util');
+const wait = promisify(setTimeout);
+
+// Test assert.rejects() and assert.doesNotReject() by checking their
+// expected output and by verifying that they do not work sync
+/* eslint-disable prefer-common-expectserror, no-restricted-properties */
+
+assert.doesNotThrow(() => {
+  assert.rejects(
+    () => assert.fail(),
+    common.expectsError({
+      code: 'ERR_ASSERTION',
+      type: assert.AssertionError,
+      message: 'Failed'
+    })
+  )
+});
+
+assert.doesNotThrow(() => assert.doesNotReject(() => {}))
+
+let promise;
+let threw = false;
+
+try {
+  promise = assert.rejects(async () => {
+    await wait(10);
+    assert.fail();
+  }, common.expectsError({
+    code: 'ERR_ASSERTION',
+    type: assert.AssertionError,
+    message: 'Failed'
+  }));
+} catch (err) {
+  threw = true;
+}
+assert.strictEqual(threw, false);
+assert.doesNotReject(() => promise);
+
+threw = false;
+try {
+  promise = assert.doesNotReject(async () => {
+    await wait(10);
+    assert.fail();
+  });
+} catch (err) {
+  threw = true;
+}
+assert.strictEqual(threw, false);
+assert.rejects(() => promise,
+  common.expectsError({
+    code: 'ERR_ASSERTION',
+    type: assert.AssertionError,
+    message: 'Got unwanted exception.\nFailed'
+  })
+);
+
+try {
+  assert.throws(() => {})
+} catch (err) {
+  const lines = err.stack.split('\n')
+  assert.equal(lines.length, 10)
+  assert.ok(lines[0].includes('Missing expected exception'))
+  assert.ok(lines[1].includes('Object.<anonymous>'))
+  assert.ok(lines[1].includes('test-assert-async.js'))
+}
+
+try {
+  assert.doesNotThrow(() => { throw new Error() })
+} catch (err) {
+  const lines = err.stack.split('\n')
+  assert.equal(lines.length, 11)
+  assert.ok(lines[0].includes('Got unwanted exception'))
+  assert.ok(lines[2].includes('Object.<anonymous>'))
+  assert.ok(lines[2].includes('test-assert-async.js'))
+}
```

Comment:
The two tests above are already tested for in `test-assert.js`. This adds the check for the error stack, so it indeed tests something new, but in that case I would rather replace the ones in `test-assert.js` with these tests instead of "duplicating" them.